### PR TITLE
Freeform puzzles

### PR DESCRIPTION
**Pull request by morganbr**
Add a concept of "freeform" puzzle answers -- puzzles that allow submitting arbitrary responses (e.g. for open-ended challenges). Authors/admins get a freeform evaluation queue where they can accept or reject submissions. Accepting a submissions marks the puzzle correct. Authors/admins can also optionally provide a response with their evaluation ("You broke rule 3" or "Nice job!") and the results will be emailed to the submitter.

### CODE REVIEW

File: ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+        private async Task PopulateUI(int? puzzleId)
+        {
+            if (puzzleId != null)
+            {
+                Puzzle = await _context.Puzzles.Where(m => m.ID == puzzleId).FirstOrDefaultAsync();
+            }
+
+            // todo: think about a way to go back to old submissions and redo them (e.g. to fix a forgotten comment)
+            // todo: should admins be able to see anything they're not an author of?
```
Comment (tabascq) [inline ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs:67]: I believe the answer to this question is yes. Admins should be able to see everything, and in a module event should be trusted to be disciplined enough not to cheat. I mean, a player-admin could just as easily look up the answer to the puzzle they need hints for, mark their own status as solved, etc. Not a blocker for tonight if it slows things down.

File: ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        }
+
+        private async Task PopulateUI(int? puzzleId)
+        {
+            if (puzzleId != null)
+            {
+                Puzzle = await _context.Puzzles.Where(m => m.ID == puzzleId).FirstOrDefaultAsync();
+            }
+
+            // todo: think about a way to go back to old submissions and redo them (e.g. to fix a forgotten comment)
```
Comment (tabascq) [inline ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs:66]: This also needs to have an audit trail, by which I only mean a way to check up on a grader and confirm that they are grading as you would expect. Also selection of greatest hits like Safari does; consider adding a "Favorite" checkbox that can be filtered later. Not a blocker for tonight if it slows things down.

File: ServerCore/Pages/Submissions/Index.cshtml.cs [hunk 3]
```diff
@@ -102,12 +104,20 @@ public async Task<IActionResult> OnPostAsync(int puzzleId, string submissionText
...
+            if (Puzzle.IsFreeform)
+            {
+                submission.UnformattedSubmissionText = submissionText;
+            }
+            else
+            {
+                submission.SubmissionText = submissionText;
+            }
+
             submission.Response = await _context.Responses.Where(
```
Comment (tabascq) [inline ServerCore/Pages/Submissions/Index.cshtml.cs:121]: This response means nothing in the freeform case, why look it up?

File: ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+                                               join PuzzleAuthors author in _context.PuzzleAuthors on sub.PuzzleID equals author.PuzzleID
+                                               where sub.ID == SubmissionID &&
+                                               sub.FreeformAccepted == null &&
+                                               author.Author == LoggedInUser
+                                               select sub).FirstOrDefaultAsync();
+
+                if (submission != null)
+                {
+                    submission.FreeformAccepted = accepted;
+                    submission.FreeformResponse = FreeformResponse;
```
Comment (tabascq) [inline ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs:110]: If FreeformResponse is null, this should just be either "Accepted" or "Rejected". Better than blank.

**Review by tabascq — APPROVED** [review]
A couple small things plus some advice for the future. Thanks!

File: ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs (hunk 4 above)
Comment (morganbr) [reply ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs:110]: Good idea. Done.

File: ServerCore/Pages/Submissions/Index.cshtml.cs (hunk 3 above)
Comment (morganbr) [reply ServerCore/Pages/Submissions/Index.cshtml.cs:121]: I had the general thought that there could be freeform puzzles that still have a single final answer like Twice from PD2020. Does that sound reasonable? (Also looking at this again, I realized I wasn't formatting freeform answers for this check; fixed that)

File: ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs (hunk 2 above)
Comment (morganbr) [reply ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs:66]: Great ideas. I'll start thinking about that for a followup. Filed #608

File: ServerCore/Pages/Submissions/Index.cshtml.cs (hunk 3 above)
Comment (tabascq) [reply ServerCore/Pages/Submissions/Index.cshtml.cs:121]: With the formatting pass, makes sense.

File: ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs (hunk 1 above)
Comment (morganbr) [reply ServerCore/Pages/Submissions/FreeformQueue.cshtml.cs:67]: Sounds reasonable, done.

**Pull request merged**